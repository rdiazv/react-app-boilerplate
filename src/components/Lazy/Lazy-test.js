import { React, mount } from 'helpers/test-helper'
import Lazy from 'components/Lazy'

const getComponent = props =>
  mount(
    <Lazy
      getComponent={() => new Promise(() => {})}
      {...props}
    />
  )

describe('Lazy', () => {
  describe('componentWillMount()', () => {
    it('should invoke loadComponent passing props.getComponent', () => {
      const callback = () => Promise.resolve({})
      const component = getComponent({
        getComponent: callback,
      })
      const instance = component.instance()

      instance.loadComponent = jest.fn()
      instance.componentWillMount()

      expect(instance.loadComponent).toHaveBeenCalledTimes(1)
      expect(instance.loadComponent).toHaveBeenCalledWith(callback)
    })
  })

  describe('componentWillReceiveProps(nextProps)', () => {
    describe('if props.getComponent changed', () => {
      it('should invoke loadComponent passing nextProps.getComponent', () => {
        const component = getComponent()
        const instance = component.instance()
        const callback = jest.fn()

        instance.loadComponent = jest.fn()
        component.setProps({
          getComponent: callback,
        })

        expect(instance.loadComponent).toHaveBeenCalledTimes(1)
        expect(instance.loadComponent).toHaveBeenCalledWith(callback)
      })
    })

    describe('if props.getComponent do not change', () => {
      it('should not invoke loadComponent', () => {
        const callback = () => Promise.resolve({})
        const component = getComponent({
          getComponent: callback,
        })
        const instance = component.instance()

        instance.loadComponent = jest.fn()
        component.setProps({
          getComponent: callback,
        })

        expect(instance.loadComponent).not.toHaveBeenCalled()
      })
    })
  })

  describe('loadComponent(getComponent)', () => {
    it('should invoke getComponent', () => {
      const component = getComponent()
      const instance = component.instance()
      const callback = jest.fn(() => new Promise(() => {}))

      instance.loadComponent(callback)

      expect(callback).toHaveBeenCalledTimes(1)
    })

    it('should assign the module default export to state.component', () => {
      const component = getComponent()
      const instance = component.instance()
      const promise = new Promise(resolve => {
        resolve({ default: 'div' })
        expect(component).toHaveState('component', 'div')
      })

      instance.loadComponent(() => promise)

      return promise
    })

    it('should set state.loading while the promise is resolving', () => {
      const component = getComponent()
      const instance = component.instance()
      const promise = new Promise(resolve => {
        setTimeout(() => resolve({ default: 'div' }), 1)
        expect(component).toHaveState('loading', false)
      })

      instance.loadComponent(() => promise)

      setTimeout(() => {
        expect(component).toHaveState('loading', true)
      })

      return promise
    })
  })

  describe('if state.loading', () => {
    it('should render props.loadingComponent', () => {
      const component = getComponent({
        loadingComponent: <div>LOADING</div>,
      })

      component.setState({ loading: true })

      expect(component).toMatchSnapshot()
    })
  })

  describe('if not state.loading and state.component', () => {
    it('should render state.component passing all extra props', () => {
      const component = getComponent({
        extra1: 'someExtraProp1',
        extra2: 'someExtraProp2',
      })

      const SomeComponent = () => <div />

      component.setState({
        loading: false,
        component: SomeComponent,
      })

      expect(component).toMatchSnapshot()
    })
  })

  describe('if not state.loading and not state.component', () => {
    it('should render nothing', () => {
      const component = getComponent()

      component.setState({
        loading: false,
        component: null,
      })

      expect(component).toMatchSnapshot()
    })
  })
})
