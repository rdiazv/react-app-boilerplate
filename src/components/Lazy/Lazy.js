import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

export default class Lazy extends PureComponent {
  static propTypes = {
    getComponent: PropTypes.func.isRequired,
    loadingComponent: PropTypes.node,
  }

  static defaultProps = {
    loadingComponent: null,
  }

  state = {
    loading: false,
    component: null,
  }

  componentWillMount() {
    this.loadComponent(this.props.getComponent)
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.getComponent !== nextProps.getComponent) {
      this.loadComponent(nextProps.getComponent)
    }
  }

  loadComponent = getComponent => {
    let resolved = false

    setTimeout(() => {
      if (!resolved) {
        this.setState({ loading: true })
      }
    })

    getComponent().then(module => {
      resolved = true

      this.setState({
        loading: false,
        component: module.default,
      })
    })
  }

  render() {
    const { loadingComponent, ...props } = this.props

    delete props.getComponent

    if (this.state.loading) {
      return loadingComponent
    }

    if (this.state.component) {
      return <this.state.component {...props} />
    }

    return null
  }
}
