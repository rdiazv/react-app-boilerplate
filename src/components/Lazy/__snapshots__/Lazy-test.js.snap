// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lazy if not state.loading and not state.component should render nothing 1`] = `
<Lazy
  getComponent={[Function]}
  loadingComponent={null}
/>
`;

exports[`Lazy if not state.loading and state.component should render state.component passing all extra props 1`] = `
<Lazy
  extra1="someExtraProp1"
  extra2="someExtraProp2"
  getComponent={[Function]}
  loadingComponent={null}
>
  <SomeComponent
    extra1="someExtraProp1"
    extra2="someExtraProp2"
  >
    <div />
  </SomeComponent>
</Lazy>
`;

exports[`Lazy if state.loading should render props.loadingComponent 1`] = `
<Lazy
  getComponent={[Function]}
  loadingComponent={
    <div>
      LOADING
    </div>
  }
>
  <div>
    LOADING
  </div>
</Lazy>
`;
